### ISSUE TITLE
Getting WinError 10054 when connecting to 127.0.0.1

### ISSUE THREAD

**Issue by pootislord** [olli-e/ISRT-Insurgency-Sandstorm-RCON-Query-Tool#4]
So the IP is correct, and the rcon and query port are the exact same as in the bat file (and I can find the rcon port being read out in the server log, but I only see the query port where it first runs the additional paramiters (i.e. where it says "Scenario=XXX?MaxPlayers=28 -Port=XXXXX -QueryPort=XXXXX -log -..." in the server log)

The exact error is:

---------------------------
ISRT Error Message
---------------------------
Something went wrong: 

[WinError 10054] An existing connection was forcibly closed by the remote host

Wrong IP or Port?
---------------------------
OK   
---------------------------

I'm able to RCON into the server using other apps, and those never asked for the query port, just IP, RCON port and RCON password.

Not sure why it's getting caught up on that specifically, any help is appreciated because I hate going and looking up the scenario names just to change the map.

Thanks.

**Comment by sargolin** [comment]
Hi pootislord, thanks for leeting me know about the error. Let's try and fix this. I know about the travel scenario struggle and I will get this running for you. 

Let me get the first idea - you did provide and rcon port, but not a query port, right? To make this clear, you would normally not need a query port in general to rcon into your sevrer. But I might have produced the error, because, when you use the map changer with the drop down menu, it calls the query script afterwards to update the running map, which will most likely runs into an error - but as I just tested this, it does only give me the "...is no valid port number..."error. 

So there's basically happening something else. Could you please give me the exakt information that you are putting in (you can mask your RCOn Port and PW of course, because I need to assume you have them correct) and what buttons or menus do you select? In addition, to be able to understand the circumstances, I need a couple of answers. 

- Are you running a local or a hosted server? 
- If local, windows or linux? 
- If local, did you try to disable the firewall for a test?
- Are you able to normally query the server? 
- Are you able to run a simple rcon command like listplayers or help?

Could you query and rcon to the server directly after another and sent me a copy of the server logs from those connections? This looks like this:

[PID 219913 Ports 27115,27116,27117] | [2020.11.09-22.53.01:771][763]LogRcon: 127.0.0.1:35958 << listplayers
[PID 219913 Ports 27115,27116,27117] | [2020.11.09-22.53.04:303][915]LogRcon: Accepted connection from 37.201.153.6:50920
[PID 219913 Ports 27115,27116,27117] | [2020.11.09-22.53.04:353][917]LogRcon: 37.201.153.6:50920 << help
[PID 219913 Ports 27115,27116,27117] | 2020/11/09 22:53:06.760 +0100 | RCON 127.0.0.1:27117 (TX >>) listplayers
[PID 219913 Ports 27115,27116,27117] | 2020/11/09 22:53:06.791 +0100 | RCON 127.0.0.1:27117 (RX <<) 

I try to figure out what is happening and what causes the error. It's not a program error, since it shows a "Windows error message with code", not my programm error messages. It tells you that you tried to connect to a specific port that already has a connection running there. So may it be that some other program connected to this port or you started ISRT twice or you had two programs connecting to this port in parallel? This would explain it - please try to check and reboot your computer and right then start ISRT and try it again.

Thanks and have a great day!
Madman

**Comment by sargolin** [comment]
Also could you please check with my Test Server if there's a difference?

IP: 93.186.198.185
Game Port: 27299
Query Port: 27201
RCON Port: 27222
RCON PW: testrconpw

Thanks,
Madmana

**Comment by sargolin** [comment]
And one last question, does the error come immediately or after a specific time? Can you tell me approximately after how many seconds? This is important, because I have a couple of time-related commands in it. Thanks m8!

**Comment by pootislord** [comment]
Okay so I'll see if I can answer everything in one go:

My batch file is this (which I run on the same PC as the game, Windows 10 64 Pro):
`InsurgencyServer.exe Scenario=Scenario_PowerPlant_Checkpoint_Security?MaxPlayers=28 -Port=27102 -QueryPort=27131 -log -hostname=XXXXXXX -Rcon -RconPassword=XXXX -RconListenPort=XXXXX`

From there I go into the application and fill in (in the box in the Top Left, and I've also added the same info into the server manager) the server IP (127.0.0.1), the Query port, the RCON port, and the RCON password. I hit enter and it immediately gives me the "Something went wrong, WinError 10054, Wrong IP or Port? etc.". If I put in another incorrect IP (124.0.0.1, for example), it waits a few seconds and says "Something went wrong, wrong IP or Port?" (I think that's the timed command you were referring to) so I know the IP is correct.

After that I can go into another RCON app, MCRCON in this case, and sign in with just the IP, RCON port, and password and that works fine.

I've not dealt with querying before so I'm not too sure to what degree that might be the issue (MCRCON didn't require it before), sorry if I'm completely missing something related to that.

I'm not exactly sure the order of how it runs the query, RCON port and Password, but changing the RCON Port, Query Port, and password to random values all give me the same error that I get when they're correct.

Also, I was unable to connect to the test server on both MCRCON and on the ISRT app. Tested just a few mins ago.

Lastly, I got a PUP notification from windows defender from the app, it said "Win32/Wacapew.C!ml". I made sure to allow it while testing the app.

Let me know if there's anything else I can test or add, and I appreciate the help!

**Comment by sargolin** [comment]
Could you test my test server again please? I had a crash last night and it is up and running again. We need to get this working before anything else, so I can be sure that the basics are fine.

IP: 93.186.198.185
Game Port: 27299
Query Port: 27201
RCON Port: 27222
RCON PW: testrconpw

**Comment by sargolin** [comment]
I analyzed the problem and was able to re-create your described behavior. I started a windows server and was not able to connect via query (which is TCP), but via RCON (which UDP traffic). my netstat showed me that the ports are actually monitored and the daemons are bound to 0.0.0.0:

```
PS C:\WINDOWS\system32> netstat -an | findstr "27141"   TCP    0.0.0.0:27141          0.0.0.0:0              
PS C:\WINDOWS\system32> netstat -an | findstr "27131"  UDP    0.0.0.0:27131          *:*
```


After I replaced my localhost address 127.0.0.1 with my actual local ipaddress from "ipconfig", I was able to connect to the system with query and everything worked perfectly. The final step is to check your c:/windows/system32/driver/etc/hosts file, if there is any re-direct for 127.0.0.1 in it. In my case, there was, which all my connects to 127.0.0.1 were re-directed to the false server:

`127.0.0.1 link.domain.com www.link.domain.com	127.0.0.1	platform.domain.com`

replaced with:

`127.0.0.1 localhost`

### To summarize:

1. Replace 127.0.0.1 with your actual network ip address from your network card (most times something like 192.168.xxx.xxx
2. Check your etc/hosts file for any redirects of 127.0.0.1
3. Give me some time to find out why 127.0.0.1 is not working, if 1 and 2 are done properly, but the app should work fine with the actual network card address.

Please let me know, if this helped for the time being. I'll try to fix the issue with 127.0.0.1
Thanks,
Madman

**Comment by sargolin** [comment]
I'll now close this issue and track it until I found a solution for the 127.0.0.1 problem - it has to do with server package response that is not correct

```
Connecting to 93.186.198.185:27201
Sending info request
Paket: b'\xff\xff\xff\xffTSource Engine Query\x00'
Received: b'\xff\xff\xff\xffI\x11ISRT Test Server\x00Canyon\x00sandstorm\x00Insurgency: Sandstorm\x00\x00\x00\x00\x02\x00dl\x00\x011.1.0.0\x00\xb1\xa3j\x10\x01:\x00\x00\x000\x01V:122064,S:523,MatchServer_b:false,ModList_s:98145,Mutators_s:NoRestrictedArea\x00\xc8\xde\x08\x00\x00\x00\x00\x00'
Single packet response
Sending rules request
Paket: b'\xff\xff\xff\xffV\xff\xff\xff\xff'
Received: b'\xff\xff\xff\xffA\xc0\x01k\n'
Single packet response
Using challenge: 174784960
Paket: b'\xff\xff\xff\xffV\xc0\x01k\n'
Received: b'\xff\xff\xff\xffE\x11\x00Coop_b\x00true\x00CustomRules_b\x00false\x00Day_b\x00true\x00GameMode_s\x00Hardcore Checkpoint\x00MatchServer_b\x00false\x00ModList_s\x0098145\x00Mods_b\x00true\x00Mutated_b\x00true\x00Mutators_s\x00NoRestrictedArea\x00Night_b\x00false\x00OfficialRuleset_b\x00false\x00PlrC_i\x000\x00PlrM_i\x002\x00Pwd_b\x00false\x00RankedServer_b\x00true\x00S\x00523\x00Versus_b\x00false\x00'
Single packet response
Connecting to 127.0.0.1:27131
Sending info request
Paket: b'\xff\xff\xff\xffTSource Engine Query\x00'
Received: b'\xff\xff\xff\xffI\x11MyTestserver\x00Farmhouse\x00sandstorm\x00Insurgency: Sandstorm\x00\x00\x00\x00 \x00dw\x00\x011.1.0.0\x00\xb1\xdei\x02\xc0\x81H\x01>@\x01V:122064,S:523,MatchServer_b:false\x00\xc8\xde\x08\x00\x00\x00\x00\x00'
Single packet response
Sending rules request
Paket: b'\xff\xff\xff\xffV\xff\xff\xff\xff'
Received: b'\xff\xff\xff\xffE\x0f\x00Coop_b\x00true\x00CustomRules_b\x00true\x00Day_b\x00true\x00GameMode_s\x00Hardcore Checkpoint\x00MatchServer_b\x00false\x00Mods_b\x00false\x00Mutated_b\x00false\x00Night_b\x00false\x00OfficialRuleset_b\x00false\x00PlrC_i\x000\x00PlrM_i\x0032\x00Pwd_b\x00false\x00RankedServer_b\x00false\x00S\x00523\x00Versus_b\x00false\x00'
Single packet response
```

### Whenever you have this problem, use the correct LAN network IP instead of localhost!

**Comment by sargolin** [comment]
For the time being I catch all 127.0.0.1 requests and make them false with the hint to use the LAN IP. That's at least a temporary fix for running into the problem in first place.

**Comment by olli-e** [comment]
Closed, revert to LAN IP automatically.
